.github/workflows/sync-with-upstream.yaml

```diff
@@ -50,34 +50,42 @@ jobs:
           && cat repo/.github/workflows/.template_version || echo "v0.0.0")
         echo "Latest version is $LATEST_VERSION and current version is $CURRENT_VERSION."
 
+        # Check if the template repo was changed/updated.
         if [ "$CURRENT_VERSION" == "$LATEST_VERSION" ]; then
           echo "Unable to find a new version, exiting..."
         else
 
-          # Copy the files from the template to the current repo.
-          rsync --recursive --verbose --verbose --exclude tests/ --exclude src/ --exclude docs/source/ \
-            --exclude CHANGELOG.md --exclude CODEOWNERS --exclude LICENSE.md --exclude README.md --exclude SECURITY.md \
-            --exclude .git/ --exclude .github/workflows/.template_version template/ repo/
+          # Check if the branch already exists in the current repo.
+          BRANCH_NAME="sync-$LATEST_VERSION"
+          if [ "$(git rev-parse --verify origin/"""$BRANCH_NAME""" 2>/dev/null)" ]; then
+            echo "Branch $BRANCH_NAME already exists, exiting..."
+          else
 
-          cd repo
+            # Generate a patch file of all template changes.
+            pushd template
+            git diff "$CURRENT_VERSION..$LATEST_VERSION" > diff.patch
+            popd
 
-          # Check if the branch exists in the current repo.
-          if [ "$(git rev-parse --verify origin/sync-"""$LATEST_VERSION""" 2>/dev/null)" ]; then
-            echo "Branch sync-$LATEST_VERSION already exists. Exiting..."
-            exit 0
-          fi
+            # Apply the generated patch to the current repo.
+            pushd repo
+            patch --strip 1 --batch --merge --input ../template/diff.patch || true
+            git checkout .github/workflows/.template_version src/ docs/ tests/ "*.md"
```

Comment:
@eduardocardoso @behnazh Instead of creating a “full” patch here

https://github.com/jenstroeger/python-package-template/blob/7a1f03fce75fde6c1a0b7807c8dbbf0bb6e3834e/.github/workflows/sync-with-upstream.yaml#L66

then using `patch` (with likely unwanted changes) and then undoing the unwanted changes using `git checkout` — why not generate a patch with only the necessary changes?

I think something like this might work:
```bash
git diff "$CURRENT_VERSION..$LATEST_VERSION" $(find . .github/workflows/ -maxdepth 1 -type f -name '*' ! -name '*.md' ! -name '.template_version') > diff.patch
```
It would also make explicit to the user which files will be part of the patch, and therefore which files are sync’ed automatically from the upstream template.

Reply:
Makes sense. There's no reason to include those files in the patch

Reply:
Please add `docs/` to the command to include `docs/Makefile`.

Reply:
Thanks @behnazh, commit https://github.com/jenstroeger/python-package-template/pull/537/commits/2aed4287fe0b7f6b968a1e8a5022150a82ae752e.